packages/ui/components/badge/Badge.tsx

```diff
@@ -41,19 +40,50 @@ type IconOrDot =
     }
   | { startIcon?: unknown; withDot?: boolean };
 
-export type BadgeProps = InferredBadgeStyles &
-  ComponentProps<"div"> & { children: ReactNode; rounded?: boolean } & IconOrDot;
+export type BadgeBaseProps = InferredBadgeStyles & {
+  children: React.ReactNode;
+  rounded?: boolean;
+} & IconOrDot;
+
+export type BadgeProps =
+  /**
+   * This union type helps TypeScript understand that there's two options for this component:
+   * Either it's a div element on which the onClick prop is not allowed, or it's a button element
+   * on which the onClick prop is required. This is because the onClick prop is used to determine
+   * whether the component should be a button or a div.
+   */
+  | (BadgeBaseProps & Omit<React.HTMLAttributes<HTMLDivElement>, "onClick"> & { onClick?: never })
+  | (BadgeBaseProps & Omit<React.HTMLAttributes<HTMLButtonElement>, "onClick"> & { onClick: () => void });
 
 export const Badge = function Badge(props: BadgeProps) {
   const { variant, className, size, startIcon, withDot, children, rounded, ...passThroughProps } = props;
+  const isButton = "onClick" in passThroughProps && passThroughProps.onClick !== undefined;
   const StartIcon = startIcon ? (startIcon as SVGComponent) : undefined;
-  return (
-    <div
-      className={classNames(badgeStyles({ variant, size }), rounded && "h-5 w-5 rounded-full p-0", className)}
-      {...passThroughProps}>
+  const classes = classNames(
+    badgeStyles({ variant, size }),
+    rounded && "h-5 w-5 rounded-full p-0",
+    className
+  );
+
+  const Children = () => (
+    <>
       {withDot ? <GoPrimitiveDot className="h-3 w-3 stroke-[3px]" /> : null}
       {StartIcon ? <StartIcon className="h-3 w-3 stroke-[3px]" /> : null}
-      <div>{children}</div>
+      {children}
+    </>
+  );
+
+  if (isButton) {
+    return (
+      <button {...passThroughProps} className={classes}>
+        <Children />
+      </button>
+    );
+  }
+
+  return (
+    <div {...passThroughProps} className={classes}>
+      <Children />
     </div>
   );
```

Comment:
We can clean this up a bit 

```suggestion
  const Wrapper  = isButton ? "button" ? "div"

  return (
    <Wrapper {...passThroughProps} className={classes}>
      <Children />
    </Wrapper>
  );
```

Reply:
We might get some type issues - not tested this

Reply:
TypeScript doesn't like that unfortunately 😭 Try it. If you can make it work you're my hero 🫶

Reply:
Fixed with this solution :) Pushed fix 

<img width="981" alt="Screenshot 2023-04-17 at 14 57 14" src="https://user-images.githubusercontent.com/55134778/232507611-32923d1e-8148-415a-9d04-49bbe7df6bb7.png">